model/bot.go

```diff
@@ -0,0 +1,210 @@
+func (b *Bot) IsValid() *AppError {
+	if !IsValidId(b.UserId) {
+		return NewAppError("Bot.IsValid", "model.bot.is_valid.user_id.app_error", b.Trace(), "", http.StatusBadRequest)
+	}
+
+	if !IsValidUsername(b.Username) {
+		return NewAppError("Bot.IsValid", "model.bot.is_valid.username.app_error", b.Trace(), "", http.StatusBadRequest)
+	}
+
+	if utf8.RuneCountInString(b.DisplayName) > BOT_DISPLAY_NAME_MAX_RUNES {
+		return NewAppError("Bot.IsValid", "model.bot.is_valid.user_id.app_error", b.Trace(), "", http.StatusBadRequest)
+	}
+
+	if utf8.RuneCountInString(b.Description) > BOT_DESCRIPTION_MAX_RUNES {
+		return NewAppError("Bot.IsValid", "model.bot.is_valid.description.app_error", b.Trace(), "", http.StatusBadRequest)
+	}
+
+	if len(b.OwnerId) == 0 || utf8.RuneCountInString(b.OwnerId) > BOT_CREATOR_ID_MAX_RUNES {
+		return NewAppError("Bot.IsValid", "model.bot.is_valid.creator_id.app_error", b.Trace(), "", http.StatusBadRequest)
+	}
+
+	if b.CreateAt == 0 {
+		return NewAppError("Bot.IsValid", "model.bot.is_valid.create_at.app_error", b.Trace(), "", http.StatusBadRequest)
+	}
+
+	if b.UpdateAt == 0 {
+		return NewAppError("Bot.IsValid", "model.bot.is_valid.update_at.app_error", b.Trace(), "", http.StatusBadRequest)
+	}
+
+	return nil
+}
+
+// PreSave should be run before saving a new bot to the database.
+func (b *Bot) PreSave() {
+	b.CreateAt = GetMillis()
+	b.UpdateAt = b.CreateAt
+	b.DeleteAt = 0
+}
+
+// PreUpdate should be run before saving an updated bot to the database.
+func (b *Bot) PreUpdate() {
+	b.UpdateAt = GetMillis()
+}
+
+// Etag generates an etag for caching.
+func (b *Bot) Etag() string {
+	return Etag(b.UserId, b.UpdateAt)
+}
+
+// ToJson serializes the bot to json.
+func (b *Bot) ToJson() []byte {
+	data, _ := json.Marshal(b)
+	return data
+}
+
+// BotFromJson deserializes a bot from json.
+func BotFromJson(data io.Reader) *Bot {
+	var bot *Bot
+	json.NewDecoder(data).Decode(&bot)
+	return bot
+}
+
+// Patch modifies an existing bot with optional fields from the given patch.
+func (b *Bot) Patch(patch *BotPatch) {
+	if patch.Username != nil {
+		b.Username = *patch.Username
+	}
+
+	if patch.DisplayName != nil {
+		b.DisplayName = *patch.DisplayName
+	}
+
+	if patch.Description != nil {
+		b.Description = *patch.Description
+	}
+}
+
+// ToJson serializes the bot patch to json.
+func (b *BotPatch) ToJson() []byte {
+	data, err := json.Marshal(b)
+	if err != nil {
+		return nil
+	}
+
+	return data
+}
+
+// BotPatchFromJson deserializes a bot patch from json.
+func BotPatchFromJson(data io.Reader) *BotPatch {
+	decoder := json.NewDecoder(data)
+	var botPatch BotPatch
+	err := decoder.Decode(&botPatch)
+	if err != nil {
+		return nil
+	}
+
+	return &botPatch
+}
+
+// UserFromBot returns a user model describing the bot fields stored in the User store.
+func UserFromBot(b *Bot) *User {
+	return &User{
+		Id:       b.UserId,
+		Username: b.Username,
+		// TODO: Allow users not to have an email.
```

Comment:
Still intending to do it, or intend to merge with the fake address?

Reply:
No. Will remove.